### PR TITLE
feat: Add scripts for parsing .out and .err from redcap-import

### PR DESCRIPTION
**Pull request by dheshanm**
Builds on top of existing 'parse_redcap_error.py' to only print files that failed with HTTP Code 400. Adds parse_redcap_errors_err.py that similarly parses '*.out' files from redcap import.

### CODE REVIEW

**Comment by tashrifbillah** [comment]
Can you remind me why you have updated `parse_redcap_error.py`, the original script?

File: parse_redcap_errors_err.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    Parses a file for errors and returns the file path and the number of errors found.
+    Expects redcap-import-*-*.err files.
+
+    Args:
+        file_path (str): The path to the file to parse.
+
+    Returns:
+        tuple: A tuple containing the file path and the number of errors found.
+    """
+    try:
```
Comment (tashrifbillah) [inline parse_redcap_errors_err.py:23]: Why do we need try/except?

File: parse_redcap_errors_err.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        with open(file_path) as file:
+            content = file.read().split('\n')
+            error_count = 0
+
+        idx = 0
+        while idx < len(content):
+            line = content[idx]
+            # If the line contains 'Traceback (most recent call last):', then there was an error
+            if 'Traceback (most recent call last):' in line:
+                error_count += 1
```
Comment (tashrifbillah) [inline parse_redcap_errors_err.py:33]: `error_count` code should be remove. This is unnecessary keeping in mind the simplicity of this program.

File: parse_redcap_errors_err.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    for file, file_error_count in results:
+        if file_error_count > 0:
+            errors_found = True
+            errors_files_count += 1
+            errors_found_count += file_error_count
+            print(f"{file}: {file_error_count} errors")
+
+    print('\n')
+    if not errors_found:
+        print('\033[92m' + 'No errors found!' + '\033[0m')
```
Comment (tashrifbillah) [inline parse_redcap_errors_err.py:81]: This log cannot be printed. Print only file name and associated errors.

File: parse_redcap_errors_err.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+    return (file_path, error_count)
+
+if __name__ == '__main__':
+    if len(sys.argv) < 2:
+        print('Usage: python {__file__} <path/and*pattern.*>')
+        sys.exit(1)
+
+    files = glob(sys.argv[1])
+    results = [parse_file(file) for file in tqdm(files)]
```
Comment (tashrifbillah) [inline parse_redcap_errors_err.py:63]: `tqdm()` is useful when program takes a while. This program, even with all 2618 enrollment will finish in minutes.

**Review by tashrifbillah — COMMENTED** [review]
At all times, keep in mind the utility of a program when enhancing it. How much impact does it have on the whole? We know that we could `cat *out > /tmp/out.txt` and then `vim /tmp/out.txt` to examine them conveniently. Instead, we are merely trying to print just the error lines. So the script for that should be quick and simple enough to write. Our successors also need to be able to manage it without great effort. Therefore, simplicity is something we need to follow for coding it.

**Comment by tashrifbillah** [comment]
Hi @dheshanm , can you review this block for me?

https://github.com/AMP-SCZ/utility/blob/b63261ca8cf92c6220cb66a35cd25fbcfe83d5c2/parse_redcap_error.py#L26-L39

I started off your idea, changed the condition for non-error lines, and omitted printing of the next CSV file line.

**Comment by dheshanm** [comment]
Great! Instead of looking for 'Traceback (most recent call last):' to identify errors, you are looking for changes in terminal color to red with the escape sequence '\033[0;31m'. 

I did not know that these escape sequences will be captured, when outputs of script is piped into a file. This is much cleaner, than what I had attempted.

**Pull request closed without merging**